sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala

```diff
@@ -832,7 +832,13 @@ private[hive] class HiveClientImpl(
     val (partCols, schema) = table.schema.map(toHiveColumn).partition { c =>
       table.partitionColumnNames.contains(c.getName)
     }
-    if (schema.isEmpty) {
+
+    // after SPARK-19279, it is not allowed to create a table with an empty schema, but we should
+    // support it when test some HiveExternalCatalog BackwardCompatibility cases to create an old
+    // spark table. the SPARK_TEST_OLD_SOURCE_TABLE_CREATE property is used to resolve this.
+    if (schema.isEmpty && (table.properties.getOrElse(
+      HiveExternalCatalog.DATASOURCE_SCHEMA_NUMPARTS, "0").toInt != 0) || table.properties
+      .getOrElse(HiveExternalCatalog.SPARK_TEST_OLD_SOURCE_TABLE_CREATE, "false").toBoolean) {
```

Comment:
but table properties should always contain `DATASOURCE_SCHEMA_NUMPARTS`, doesn't it?